File: readers_writer/reader_writer.py
from threading import Thread, Semaphore, Lock
from random import randint
from time import sleep

#semaphore = Semaphore(1)
lock = Lock()

buffer = [0,1]

def writer():
    sleep(randint(1,4))
    #semaphore.acquire()
    lock.acquire()
    buffer[0] += 1
    buffer[1] += 1
    print("Wrote to buffer, buffer[0]:", buffer[0],"buffer[1]:",buffer[1])
    #semaphore.release()
    lock.release()

def reader(id):
    sleep(randint(1,4))
    #semaphore.acquire()
    lock.acquire()
    print("Read buffer[0]:", buffer[0]," buffer[1]:", buffer[1])
    #semaphore.release()
    lock.release()

if __name__ == "__main__":
    threads = []
    for i in range(5):
        for j in range(3):
            r = Thread(target=reader, args=(j,))
            threads.append(r)
        w = Thread(target=writer)
        threads.append(w)
    for t in threads:
        t.start()
    for t in threads:
        t.join()

File: readers_writer/reader_writer.go
package main

import (
    "fmt"
    "time"
    "math/rand"
)

var channel = make(chan int, 1)
var buffer = []int{0,1}
var done = make(chan bool)

func writer() {
    time.Sleep(time.Duration(rand.Intn(4-1)+1)*time.Second)
    <-channel   // Lock
    buffer[0] += 1
    buffer[1] += 1
    fmt.Println("Wrote to buffer, buffer[0]:", buffer[0],"buffer[1]:",buffer[1])
    channel <-1 // Unlock
    done <-true
}

func reader() {
    time.Sleep(time.Duration(rand.Intn(4-1)+1)*time.Second)
    <-channel   // Lock
    fmt.Println("Read buffer[0]:", buffer[0]," buffer[1]:", buffer[1])
    channel <-1 // Unlock
    done <-true
}

func main() {

    fmt.Println("lol")
    channel <-1

    for i := 0; i < 5; i++ {
        go reader()
        go reader()
        go reader()
        go writer()
    }
    for i := 0; i < 5; i++ {
        <-done
        <-done
        <-done
        <-done
    }
}
